consensus/src/block/free.rs

```diff
@@ -7,6 +7,35 @@ use cuprate_types::TransactionVerificationData;
 
 use crate::ExtendedConsensusError;
 
+/// Orders the [`TransactionVerificationData`] list the same as it appears in [`Block::transactions`]
+pub(crate) fn order_transactions(
+    block: &Block,
+    txs: &mut [TransactionVerificationData],
+) -> Result<(), ExtendedConsensusError> {
+    if block.transactions.len() != txs.len() {
+        return Err(ExtendedConsensusError::TxsIncludedWithBlockIncorrect);
+    }
+
+    for (i, tx_hash) in block.transactions.iter().enumerate() {
+        if tx_hash != &txs[i].tx_hash {
+            let at_index = txs[i..]
+                .iter()
+                .position(|tx| &tx.tx_hash == tx_hash)
+                .ok_or(ExtendedConsensusError::TxsIncludedWithBlockIncorrect)?;
+
+            txs.swap(i, at_index);
+        }
+    }
```

Comment:
This ordering algo seems to error:
```rust
//! ```Cargo.toml
//! [package]
//! name = "order"
//! version = "0.1.0"
//! edition = "2021"
//!
//! [dependencies]
//! rand = "*"
//! ```

use rand::{seq::SliceRandom, thread_rng};

fn main() {
    // generate 100 transaction hashes
    let v1: Vec<[u8; 32]> = (0..100).map(|_| rand::random()).collect();

    // create a clone, but ordered differently
    let mut v2 = v1.clone();
    v2.shuffle(&mut thread_rng());

    // both vecs contain the same hashes but are in different ordering
    assert_eq!(v1.len(), v2.len());
    assert!(v1 != v2);

    // ordering function errors
    assert!(order_transactions(v1, v2).is_err());
}

fn order_transactions(v1: Vec<[u8; 32]>, mut v2: Vec<[u8; 32]>) -> Result<(), ()> {
    for (i, tx_hash) in v1.iter().enumerate() {
        if tx_hash != &v2[i] {
            let Some(at_index) = v2[i..].iter().position(|tx| tx == tx_hash) else {
                return Err(());
            };

            v2.swap(i, at_index);
        }
    }

    Ok(())
}
```

Reply:
Also here are some interesting perf tests:
```
full_clone: 0.12496812
iter_branch_clone: 0.15211749
iter_position_swap: 2.227751
```
```rust
//! ```Cargo.toml
//! [package]
//! name = "order"
//! version = "0.1.0"
//! edition = "2021"
//!
//! [dependencies]
//! rand = "*"
//! ```

use std::{hint::black_box as b, time::Instant};

use rand::{seq::SliceRandom, thread_rng};

fn main() {
    let v1: Vec<[u8; 32]> = (0..100).map(|_| rand::random()).collect();
    let mut v2: Vec<[u8; 32]> = v1.clone();
    v2.shuffle(&mut thread_rng());

    assert_eq!(v1.len(), v2.len());
    assert!(v1 != v2);

    // clone entire vec if `!=`
    test(b(v1.clone()), b(v2.clone()), full_clone, "full_clone");

    // iter, clone hash if `!=`
    test(
        b(v1.clone()),
        b(v2.clone()),
        iter_branch_clone,
        "iter_branch_clone",
    );

    // same algo as PR but ignores errors
    test(
        b(v1.clone()),
        b(v2.clone()),
        iter_position_swap,
        "iter_position_swap",
    );
}

fn test(
    v1: Vec<[u8; 32]>,
    v2: Vec<[u8; 32]>,
    f: fn(v1: &Vec<[u8; 32]>, v2: &mut Vec<[u8; 32]>),
    name: &str,
) {
    let now = Instant::now();
    for _ in 0..1000000 {
        f(&b(v1.clone()), &mut b(v2.clone()));
    }
    println!("{name}: {}", now.elapsed().as_secs_f32());
}

fn full_clone(v1: &Vec<[u8; 32]>, v2: &mut Vec<[u8; 32]>) {
    if v1 != v2 {
        *v2 = v1.clone();
    }
}

fn iter_branch_clone(v1: &Vec<[u8; 32]>, v2: &mut Vec<[u8; 32]>) {
    for (h1, h2) in v1.iter().zip(v2) {
        if h1 != h2 {
            *h2 = *h1;
        }
    }
}

fn iter_position_swap(v1: &Vec<[u8; 32]>, v2: &mut Vec<[u8; 32]>) {
    for (i, tx_hash) in v1.iter().enumerate() {
        if tx_hash != &v2[i] {
            if let Some(at_index) = v2[i..].iter().position(|tx| tx == tx_hash) {
                v2.swap(i, at_index);
            }
        }
    }
}
```

Reply:
Damn nice find, should have added a test. FWIW peers send txs in order which is why this worked fine, obviously we shouldn't rely on peers doing that though.

The error is so subtle:

```suggestion
    for (i, tx_hash) in block.transactions.iter().enumerate() {
        if tx_hash != &txs[i].tx_hash {
            let at_index = txs[i..]
                .iter()
                .position(|tx| &tx.tx_hash == tx_hash)
                .ok_or(ExtendedConsensusError::TxsIncludedWithBlockIncorrect)?;

            // The above `position` will give an index from inside its view of the slice so we need to add the difference.
            txs.swap(i, (i + at_index);
        }
    }
```

Reply:
FWIW the thing we are swapping are `Transaction`s which would be expensive to clone